### ISSUE TITLE
ALS external view taxi or landing lights

### ISSUE THREAD

**Issue by wlbragg** [c172p-team/c172p-detailed#811]
This is the instructions for adding the external ALS runway lights. If anyone wants to tackle this before I do, have at it. If not I'll try to get to it in the next week.

> If you're interested in this, try adapting the manager script below for the aircraft of your choice (should give decent results for the C-172p with main landing light and two position lights rendered onto the runway). There's support for up to five lightspots in the shader - just extend to get the remaining two. Though only the first two slots can be used to draw elongated shapes...
> 
> The biggest issue right now is that there's a delay between viewpoint coordinates appearing in the property tree and viewpoint coordinates affecting rendering - which (like for the regular secondary lights) gives you a momentary offset of the lights if you pan around at low framerate.
> 
> I suppose this needs to be fixed FG-side somehow - it's not a Nasal issue, property rules used for the purpose suffer from exactly the same problem.
> 
> Otherwise the script should be fairly straightforward. Call init() to initialize the values and start it, call lightN_on to switch light N on, and so on. The projection of the landing light is pretty simplistic - with some trigonometry it ought to be easy to do better, but I didn't want to overdesign at this point.

``` javascript
# provides relative vectors from eye-point to aircraft lights
# in east/north/up coordinates the renderer uses

var light_manager = {
    run: 0,

    lat_to_m: 110952.0,
    lon_to_m: 0.0,

    light1_xpos: 0.0,
    light1_ypos: 0.0,
    light1_zpos: 0.0,
    light1_r: 0.0,
    light1_g: 0.0,
    light1_b: 0.0,
    light1_size: 0.0,
    light1_stretch: 0.0,

    light2_xpos: 0.0,
    light2_ypos: 0.0,
    light2_zpos: 0.0,
    light2_r: 0.0,
    light2_g: 0.0,
    light2_b: 0.0,
    light2_size: 0.0,

    light3_xpos: 0.0,
    light3_ypos: 0.0,
    light3_zpos: 0.0,
    light3_r: 0.0,
    light3_g: 0.0,
    light3_b: 0.0,
    light3_size: 0.0,

    init: func {
        # define your lights here

        # light 1 ########
        # offsets to aircraft center
        me.light1_xpos = 20.0;
        me.light1_ypos = 4.0;
        me.light1_zpos = 2.0;

        # color values
        me.light1_r = 0.5;
        me.light1_g = 0.5;
        me.light1_b = 0.5;

        # spot size
        me.light1_size = 12.0;
        me.light1_stretch = 1.0;

        # light 2 ########
        # offsets to aircraft center
        me.light2_ypos = -6.0;
        me.light2_zpos = 2.0;

        # color values
        me.light2_r = 0.0;
        me.light2_g = 0.2;
        me.light2_b = 0.0;

        # spot size
         me.light2_size = 6.0;

        # light 3 ########
        # offsets to aircraft center
        me.light3_ypos = 6.0;
        me.light3_zpos = 2.0;

        # color values
        me.light3_r = 0.2;
        me.light3_g = 0.0;
        me.light3_b = 0.0;

        # spot size
        me.light3_size = 6.0;

        me.start();
    },

    start: func {
        setprop("/sim/rendering/als-secondary-lights/num-lightspots", 3);

        setprop("/sim/rendering/als-secondary-lights/lightspot/size", me.light1_size);
        setprop("/sim/rendering/als-secondary-lights/lightspot/size[1]", me.light2_size);
        setprop("/sim/rendering/als-secondary-lights/lightspot/size[2]", me.light3_size);

        setprop("/sim/rendering/als-secondary-lights/lightspot/stretch", me.light1_stretch);

        me.run = 1;       
        me.update();
    },

    stop: func {
        me.run = 0;
    },

    update: func {
        if (me.run == 0) {
            return;
        }

        var apos = geo.aircraft_position();
        var vpos = geo.viewer_position();

        me.lon_to_m = math.cos(apos.lat()*math.pi/180.0) * me.lat_to_m;

        var heading = getprop("/orientation/heading-deg") * math.pi/180.0;

        var lat = apos.lat();
        var lon = apos.lon();
        var alt = apos.alt();

        var sh = math.sin(heading);
        var ch = math.cos(heading);

        # light 1 position
        var alt_agl = getprop("/position/altitude-agl-ft");

        var proj_x = alt_agl;
        var proj_z = alt_agl/10.0;

        apos.set_lat(lat + ((me.light1_xpos + proj_x) * ch + me.light1_ypos * sh) / me.lat_to_m);
        apos.set_lon(lon + ((me.light1_xpos + proj_x)* sh - me.light1_ypos * ch) / me.lon_to_m);

        var delta_x = (apos.lat() - vpos.lat()) * me.lat_to_m;
        var delta_y = -(apos.lon() - vpos.lon()) * me.lon_to_m;
        var delta_z = apos.alt()- proj_z - vpos.alt();

        setprop("/sim/rendering/als-secondary-lights/lightspot/eyerel-x-m", delta_x);
        setprop("/sim/rendering/als-secondary-lights/lightspot/eyerel-y-m", delta_y);
        setprop("/sim/rendering/als-secondary-lights/lightspot/eyerel-z-m", delta_z);
        setprop("/sim/rendering/als-secondary-lights/lightspot/dir", heading);

        # light 2 position
        apos.set_lat(lat + (me.light2_xpos * ch + me.light2_ypos * sh) / me.lat_to_m);
        apos.set_lon(lon + (me.light2_xpos * sh - me.light2_ypos * ch) / me.lon_to_m);

        delta_x = (apos.lat() - vpos.lat()) * me.lat_to_m;
        delta_y = -(apos.lon() - vpos.lon()) * me.lon_to_m;
        delta_z = apos.alt() - vpos.alt();

        setprop("/sim/rendering/als-secondary-lights/lightspot/eyerel-x-m[1]", delta_x);
        setprop("/sim/rendering/als-secondary-lights/lightspot/eyerel-y-m[1]", delta_y);
        setprop("/sim/rendering/als-secondary-lights/lightspot/eyerel-z-m[1]", delta_z);

        # light 3 position
        apos.set_lat(lat + (me.light3_xpos * ch + me.light3_ypos * sh) / me.lat_to_m);
        apos.set_lon(lon + (me.light3_xpos * sh - me.light3_ypos * ch) / me.lon_to_m);

        delta_x = (apos.lat() - vpos.lat()) * me.lat_to_m;
        delta_y = -(apos.lon() - vpos.lon()) * me.lon_to_m;
        delta_z = apos.alt() - vpos.alt();

        setprop("/sim/rendering/als-secondary-lights/lightspot/eyerel-x-m[2]", delta_x);
        setprop("/sim/rendering/als-secondary-lights/lightspot/eyerel-y-m[2]", delta_y);
        setprop("/sim/rendering/als-secondary-lights/lightspot/eyerel-z-m[2]", delta_z);

        settimer(func me.update(), 0.0);
    },

    light1_on: func {
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-r", me.light1_r);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-g", me.light1_g);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-b", me.light1_b);
    },

    light1_off: func {
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-r", 0.0);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-g", 0.0);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-b", 0.0);
    },

    light2_on: func {
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-r[1]", me.light2_r);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-g[1]", me.light2_g);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-b[1]", me.light2_b);
    },

    light2_off: func {
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-r[1]", 0.0);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-g[1]", 0.0);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-b[1]", 0.0);
    },

    light3_on: func {
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-r[2]", me.light3_r);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-g[2]", me.light3_g);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-b[2]", me.light3_b);
    },

    light3_off: func {
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-r[2]", 0.0);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-g[2]", 0.0);
        setprop("/sim/rendering/als-secondary-lights/lightspot/lightspot-b[2]", 0.0);
    },

}; 
```

**Comment by onox** [comment]
This code should use the `UpdateLoop` instead of the raw `settimer()` calls. See `TiedownPositionUpdater` in `Nasal/tiedowns.nas` for an example.

**Comment by onox** [comment]
Merged in PR #843
